diff --git a/addons/ttf/ttf.c b/addons/ttf/ttf.c
index b8e0325..b935118 100644
--- a/addons/ttf/ttf.c
+++ b/addons/ttf/ttf.c
@@ -249,7 +249,7 @@ static unsigned char *alloc_glyph_region(ALLEGRO_TTF_FONT_DATA *data,
       relock = true;
    }
 
-   if (relock) {
+   if (new || data->page_lr == NULL) {
       char *ptr;
       int i;
       unlock_current_page(data);
@@ -268,22 +268,21 @@ static unsigned char *alloc_glyph_region(ALLEGRO_TTF_FONT_DATA *data,
          data->lock_rect.h = h4;
       }
 
-      data->page_lr = al_lock_bitmap_region(page,
-         data->lock_rect.x, data->lock_rect.y,
-         data->lock_rect.w, data->lock_rect.h,
-         ALLEGRO_PIXEL_FORMAT_ABGR_8888_LE, ALLEGRO_LOCK_WRITEONLY);
+      data->page_lr = al_lock_bitmap(page, ALLEGRO_PIXEL_FORMAT_ABGR_8888_LE, ALLEGRO_LOCK_READWRITE);
 
       if (!data->page_lr) {
          return NULL;
       }
 
-      /* Clear the data so we don't get garbage when using filtering
-       * FIXME We could clear just the border but I'm not convinced that
-       * would be faster (yet)
-       */
-      for (i = 0; i < data->lock_rect.h; i++) {
-          ptr = (char *)(data->page_lr->data) + (i * data->page_lr->pitch);
-          memset(ptr, 0, data->lock_rect.w * 4);
+      if (new) {
+         ptr = data->page_lr->data;
+         int n = data->page_lr->pitch * al_get_bitmap_height(page);
+         if (n < 0) {
+            n -= data->page_lr->pitch;
+            n = -n;
+            ptr -= n;
+         }
+         memset(ptr, 0, n);
       }
    }
 
@@ -291,8 +290,8 @@ static unsigned char *alloc_glyph_region(ALLEGRO_TTF_FONT_DATA *data,
 
    /* Copy a displaced pointer for the glyph. */
    return (unsigned char *)data->page_lr->data
-      + ((glyph->region.y + 1) - data->lock_rect.y) * data->page_lr->pitch
-      + ((glyph->region.x + 1) - data->lock_rect.x) * sizeof(int32_t);
+      + (glyph->region.y + 1) * data->page_lr->pitch
+      + (glyph->region.x + 1) * sizeof(int32_t);
 }
 
 
